### PR TITLE
Exclude children of subdirectories from symlink

### LINKED ISSUE #1029: Input directory corruption and infinite symlink using successive InitialWorkDirRequirement
There appear to be 2 bugs in `cwltool` handling of `InitialWorkDirRequirement`, in the use of symlinks to optimize and avoid making copies of directories:
- Directories get corrupted with incorrect symlinks that point to themselves.
- Symlinks that point to themselves cause an inifinite loop in `cwltool`

The issue is subtle, and seems to manifest when different parts of a single directory structure are referenced.

## Expected Behavior
Input directories should be left pristine, and output directories should be emitted correctly, when using `InitialWorkDirRequirement`. In the example workflow, a subdirectory with one test file should be the final output.

## Actual Behavior
A symlink is installed at `dir/subdir/file` with target being itself (1st bug), causing an infinite loop in `cwltool` as it tries to dereference the symlink within `PathMapper.visit()` (2nd bug).

## Workflow Code
Here is a code example that demonstrates the issue: `test.cwl`
```
#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: Workflow

inputs: []

steps:
  # Create a test directory structure; could be done outside CWL and passed in as input.
  # This input directory should be left pristine.
  mkdirs:
    run:
      class: CommandLineTool
      baseCommand: [bash, '-c', 'mkdir dir dir/subdir && touch dir/subdir/file', '-']
      inputs: []
      outputs:
        mkdirs_out:
          type: Directory
          outputBinding:
            glob: dir
    in: []
    out: [mkdirs_out]

  # Given an input directory, emit a subdirectory as output.
  passthrough1:
    run:
      class: CommandLineTool
      requirements:
      - class: InitialWorkDirRequirement
        listing:
        - entry: $(inputs.passthrough1_in)
          writable: false
      baseCommand: ["true"]
      inputs:
        passthrough1_in:
          type: Directory
      outputs:
        passthrough1_subdir:
          type: Directory
          outputBinding:
            glob: $(inputs.passthrough1_in.basename)/subdir
    in:
      passthrough1_in: mkdirs/mkdirs_out
    out: [passthrough1_subdir]

  # Given a (sub-)directory, emit it unchanged.
  passthrough2:
    run:
      class: CommandLineTool
      requirements:
      - class: InitialWorkDirRequirement
        listing:
        - entry: $(inputs.passthrough2_in)
          writable: false
      baseCommand: ["true"]
      inputs:
        passthrough2_in:
          type: Directory
      outputs:
        passthrough2_subdir:
          type: Directory
          outputBinding:
            glob: $(inputs.passthrough2_in.basename)
    in:
      passthrough2_in: passthrough1/passthrough1_subdir
    out: [passthrough2_subdir]

outputs:
  out:
    type: Directory
    outputSource: passthrough2/passthrough2_subdir
```

As input, use this empty `test.yaml`:
```
{}
```
## Full Traceback
Infinite loop in `pathmapper.py` around line 249, in `PathMapper.visit()`; see the comment line `# Dereference symbolic links`.

## Your Environment
* cwltool version: 1.0.20181217162649

### PR DESCRIPTION
**Pull request by jfennick**
This PR (hopefully) fixes #1029.

### CODE REVIEW

**Comment by mr-c** [comment]
Thanks for this! The lint error is being fixed over in https://github.com/common-workflow-language/cwltool/pull/1691

**Comment by jfennick** [comment]
It looks like my change is breaking some other examples, but I'm a bit lost. Is anyone willing to take a look and help push this PR across the finish line?

**Comment by mr-c** [comment]
Here is one of the conformance tests, that now fail

```
cwltool v1.0/cat-from-dir.cwl v1.0/cat-from-dir-with-literal-file.yaml
Pipe to stdin from literal File via a Directory literal
Returned non-zero
ERROR'cat' not found: [Errno 2] No such file or directory: "I'm a File literal; howdy!"
WARNING[job cat-from-dir.cwl] completed permanentFail
WARNINGFinal process status is permanentFail
```

**Comment by jfennick** [comment]
Yes, I read through the logs. I was asking if any of the maintainers are willing to help fix the errors.

File: cwltool/pathmapper.py [hunk 1]
```diff
@@ -191,6 +192,17 @@ def files(self) -> List[str]:
     def items(self) -> List[Tuple[str, MapperEnt]]:
         return list(self._pathmap.items())
 
+    def items_exclude_children(self) -> List[Tuple[str, MapperEnt]]:
+        newitems = {}
+        keys = [key for key, entry in self.items()]
+        for key, entry in self.items():
+            if entry.type in ("File", "Directory"):
```
Comment (kinow) [inline cwltool/pathmapper.py:199]: @jfennick I think this line can be removed. Debugging the conformance test that @mr-c pointed is failing, I noticed that the `File` to be staged by the workflow is a `CreateFile` (which I was not familiar with, but the code comments help clarify it, I think - <https://github.com/common-workflow-language/cwltool/blob/3945a18c699c3ca500a9343c40b34db966c6787b/cwltool/pathmapper.py#L48-L49>).

After I removed that line, the test `cat-from-dir.cwl` passed, and the workflow `test.cwl` in #1029 also passed with no errors (I switched to `main` and confirmed it failed for me too there).

**Review by kinow — COMMENTED** [review]
Debugged the test @mr-c linked, and suggested a possible fix for @jfennick.

@jfennick maybe we should include the test from #1029 as a unit test here?

Hope that helps!
-Bruno

**Review by mr-c — CHANGES_REQUESTED** [review]
Looks great! It will need a confirmatory test.

**Comment by mr-c** [comment]
Try `make install-dep format`

**Review by mr-c — APPROVED** [review]

File: cwltool/pathmapper.py [hunk 2]
```diff
@@ -191,6 +192,16 @@ def files(self) -> List[str]:
     def items(self) -> List[Tuple[str, MapperEnt]]:
         return list(self._pathmap.items())
 
+    def items_exclude_children(self) -> List[Tuple[str, MapperEnt]]:
```
Comment (mr-c) [inline cwltool/pathmapper.py:195]: FYI @tetron @DailyDreaming @kellrott  about this proposed addition to PathMapper

**Comment by jfennick** [comment]
> Try `make install-dep format`

Sorry for blasting your CI for trivial formatting, but I was running into nasty installation errors when trying to run it locally using the instructions in the readme. If I can reproduce, I'll try to open another issue.

**Comment by mr-c** [comment]
> > Try `make install-dep format`
> 
> Sorry for blasting your CI for trivial formatting, but I was running into nasty installation errors when trying to run it locally using the instructions in the readme. If I can reproduce, I'll try to open another issue.

S'alright. New commits will cancel old CI jobs automatically.

Did you have a dedicated Python virtualenv active before running those commands?

**Review by kinow — APPROVED** [review]

**Comment by jfennick** [comment]
> > > Try `make install-dep format`
> > 
> > 
> > Sorry for blasting your CI for trivial formatting, but I was running into nasty installation errors when trying to run it locally using the instructions in the readme. If I can reproduce, I'll try to open another issue.
> 
> S'alright. New commits will cancel old CI jobs automatically.
> 
> Did you have a dedicated Python virtualenv active before running those commands?

I created a new conda environment. See #1694

**Pull request merged**